Lock migrations to support multiple concurrent migrators

blatyo opened the pull request:
This is still a work in progress. I wanted to put this up for feedback.

* Locks migrations table in a transaction
* Wraps each migration in a Task, which may have its own transaction
* Propogates errors up, while ensure transactions are committed

josevalim commented on lib/ecto/migrator.ex:
```diff
@@ -105,13 +120,27 @@ defmodule Ecto.Migrator do
+    end)
+    |> Task.await(:infinity)
+    |> case do
+      {:error, error, stacktrace} ->
```
A task would never return this. Instead it would cause the linked process to fail.

josevalim commented on lib/ecto/migrator.ex:
```diff
@@ -105,13 +120,27 @@ defmodule Ecto.Migrator do
+    Task.async(fn ->
+      do_run_maybe_in_transaction(repo, module, fun)
+    end)
+    |> Task.await(:infinity)
```
It is better to use `Task.yield` instead of rescuing inside the Task but you would probably need a Task.Supervisor so you can use `async_nolink`.

josevalim replied:
Or alternatively we just let the task crash and bring the link down. The only concern is if the error report will be too verbose or not.

josevalim replied:
Let's go with let it crash first and then we can review the logged output after this is merged. :)

josevalim commented on lib/ecto/migrator.ex:
```diff
@@ -177,14 +214,36 @@ defmodule Ecto.Migrator do
+      |> Map.put(:prefix, opts[:prefix])
+      |> repo.all(timeout: :infinity, log: false)
+
+      try do
```
It seems this try catch is also not necessary.

blatyo replied:
The reason for this try catch is so that if an exception occurs on the 5th migration, we still commit the schema migrations transaction, which records that we did the first 4.

josevalim commented:
Thanks @blatyo!

I think we can directly change the migrated_versions function to include the lock:

```elixir
  def migrated_versions(repo, prefix) do
    from(p in {get_source(repo), __MODULE__}, select: p.version)
    |> Map.put(:lock, ^repo.__adapter__.lock_for_migration)
    |> Map.put(:prefix, prefix)
    |> repo.all(@opts)
  end
```

This way all the function the adapter function simply returns the lock string and the `lock_for_migration` function only needs to start the transaction.

josevalim commented:
I see. I am not a fan of the excessive use of rescue and raise. How tricky
would be to change the underlying migration code to return :ok | {:error,
_} some we just reraise at the very end if necessary?
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

blatyo commented:
Is it expected that adapters are suppose to be able to understand the lock piece of the query and turn that into this, for databases like db2, derby, and oracle?

``` sql
LOCK TABLE schema_migrations IN EXCLUSIVE MODE;
SELECT version FROM schema_migrations;
```

I was assuming they weren't, which is why I was doing the lock piece separately. If we're only worried about existing adapters, I believe what you suggest is possible with `lock: "WITH (TABLOCKX)"` for the MSSQL adapter.

It also seems like the mnesia adapter could theoretically support this with [lock/2](http://erlang.org/doc/man/mnesia.html#lock-2).

josevalim commented:
@blatyo  a good point, so we should probably let the adapter do the whole query, including starting the parent transaction?

blatyo commented:
@josevalim, I have the same hesitance with `try`/`rescue`. I think switching to `:ok` / `{:error, reason}` style is possible for all the `raise`s that the migrator produces. However, if an exception happens in the user code, we'll still need to rescue that and convert it to an error tuple. Also, since the error is in the user's code, we'll want to preserve the original error and stacktrace.

As long as `raise`s aren't used in the migrator itself, I think it would be fine to do the transaction and lock in the adapter.

josevalim commented:
Sounds good. Then we should use `catch kind:reason`, store it as `{kind, reason, System.stacktrace}` and then re-raise it with `:erlang.raise/3`.

blatyo commented:
I implemented the adapter and changed the error propagation. 

Is what I did for the adapter ok? It felt like maybe I was doing something dirty by using something at a higher level abstraction than what I was implementing. Maybe there's a reasonable way to call `execute/6` directly, but I didn't really understand what its arguments were.

For errors that were raised by the migrator, I instead had it return `{:error, exception}` and that will get raised outside the `lock_for_migration`. User error is caught as `{kind, reason, stacktrace}` and re-raised with `:erlang.raise/3`.

josevalim commented on lib/ecto/adapters/sql.ex:
```diff
@@ -596,6 +602,23 @@ defmodule Ecto.Adapters.SQL do
+    import Ecto.Query, only: [from: 2]
+
+    transaction(repo, opts, fn ->
+      from(p in {SchemaMigration.get_source(repo), SchemaMigration}, select: p.version)
```
I don't like this dependency on schema migrations, it feels a bit backwards. Maybe we can pass the initial query as argument to `lock_for_migrations`? Also, I think we don't even need the `SchemaMigration` name, we can probably write it as:

    from(p in SchemaMigration.get_source(repo), select: type(p.version, :integer))

Other than that, the callback looks great! :+1:

blatyo replied:
So, I could pass the query, but not every adapter would use the same query. If I instead passed the table name, is there a form of calling `execute/6`, that would work? The actually query for the adapters is static aside from the table name.

josevalim commented on lib/ecto/adapters/sql.ex:
```diff
@@ -596,6 +602,23 @@ defmodule Ecto.Adapters.SQL do
+      |> Map.put(:lock, "FOR UPDATE")
+      |> repo.all(timeout: :infinity, log: false)
+
+      fun.()
```
I was thinking we could call this `fun.()` with the versions as argument. What do you think?

blatyo replied:
No problem with that. It'll reduce some of the duplication in the functions passed to this.

josevalim commented on test/support/test_repo.exs:
```diff
@@ -108,6 +108,11 @@ defmodule Ecto.TestAdapter do
 
+  def lock_for_migrations(repo, opts, fun) do
+    send self(), {:lock_for_migrations, fun}
+    {:ok, fun.()}
```
If we call `fun.()` with the migration versions, we can probably get rid of [this hack](https://github.com/blatyo/ecto/blob/4808d1381a61e42c64e4cb5b798fb1f0e271bc16/test/support/test_repo.exs#L38-L41) and move it down here - which is a bit more appropriate.

michalmuskala commented:
How does it interact with adapters that don't support transactions?

josevalim commented:
@michalmuskala they will have to implement lock_for_migrations accordingly.

blatyo commented:
@josevalim I added table as an argument to the callback and moved the lookup of migration versions into the adapter. I also completely removed `SchemaMigration.migrated_versions/2`, since that work moved completely into the adapter. After I did that, I was able to remove the hack you mentioned.

Now that the calls to `SchemaMigration` are removed does the callback look good to you?

josevalim commented on lib/ecto/adapter/migration.ex:
```diff
@@ -50,4 +50,7 @@ defmodule Ecto.Adapter.Migration  do
   @callback execute_ddl(repo :: Ecto.Repo.t, command, options :: Keyword.t) :: :ok | no_return
+
+  @callback lock_for_migrations(repo :: Ecto.Repo.t, table :: String.t, options :: Keyword.t, function :: fun) ::
+    {:ok, any} | {:error, any}
```
Here you say it can return `{:error, _}` but when you call it in migration you don't check for the error tuple. Should we always say it returns `{:ok, any}` then?

josevalim replied:
We could also say it simply returns `any`. i.e. we don't expect the operation to fail. Or it returns the value or it raises.

josevalim commented on lib/ecto/migrator.ex:
```diff
@@ -42,8 +42,9 @@ defmodule Ecto.Migrator do
   def migrated_versions(repo, opts \\ []) do
     verbose_schema_migration repo, "retrieve migrated versions", fn ->
       SchemaMigration.ensure_schema_migrations_table!(repo, opts[:prefix])
-      SchemaMigration.migrated_versions(repo, opts[:prefix])
```
You have removed the `lock_for_migrations` outside of the `verbose_schema_migration` function but I think it should still be inside. Otherwise we lose the helpful error message that is part of `verbose_schema_migration` - which is meant to add folks with a bad schema_migration setup.

josevalim commented on lib/ecto/adapters/sql.ex:
```diff
@@ -596,6 +602,23 @@ defmodule Ecto.Adapters.SQL do
+        from(p in table, select: type(p.version, :integer))
+        |> Map.put(:prefix, opts[:prefix])
+        |> Map.put(:lock, "FOR UPDATE")
+        |> repo.all()
```
What happens if `repo.all` fails? How will this error cascade and show up to users? You can for example use "FOR WHATEVER" as the lock and see how it will behave.

josevalim replied:
To clarify, an error can happen if the user already has a schema_migrations table but it doesn't have a version field or something of sorts.

josevalim commented:
Thank you @blatyo. I have added two last comments and we should be good to go!

josevalim commented:
Btw, the build is failing and also please take a look at these warnings: https://travis-ci.org/elixir-ecto/ecto/jobs/274429724#L813-L820

blatyo commented:
Both points are valid. I'll need to look at both a little longer to decide.

I've been watching the build, but didn't want to go far into changing the tests without knowing where the design would go. I've started to fix them. One thing I've noticed is that the tests assume that the migrations and transactions are running in the same process as the test. That's particularly problematic now that the migrations run within a `Task`. I'm not quite sure what to do for [this test](https://github.com/elixir-ecto/ecto/blob/master/test/ecto/migrator_test.exs#L298-L304).

Finally, in the test output I notice a log of error logs like the following. Would you accept a separate PR to capture logs in the tests so that these don't show up in the test output?

```
** (Mariaex.Error) (1049): Unknown database 'ecto_test'
    (db_connection) lib/db_connection/connection.ex:148: DBConnection.Connection.connect/2
    (connection) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: nil
```

josevalim commented:
> Finally, in the test output I notice a log of error logs like the following. Would you accept a separate PR to capture logs in the tests so that these don't show up in the test output?

Yes, definitely.

josevalim commented:
> I'm not quite sure what to do for this test.

I am not sure either. You can comment the problematic tests and I can take a look at them as well before merging. :D

josevalim commented:
Thank you @blatyo, I have merged this! :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

I have done the following changes before merging:

1. I have removed part of the alter.exs tests the parent migration was causing the whole transaction to lock due to the migration locks

2. I have changed the lock_for_migrations contract to receive a query back again. The rationale is that, even for things such as the ORACLE query you posted, you can still generate the SQL and then use it as two separated commands.

If you have more questions or more improvements, please let us know or send more PRs. Thanks!

blatyo commented:
Thanks for taking it across the finish line! Also, thanks for all the feedback! 

I suppose for other adapters you can introspect the query passed to get the options to generate the locking query and that should be sufficient.

josevalim commented:
@blatyo yes, the Ecto.Query construct is part of the adapter contract, so they can do that. Getting the table name is a matter of calling query.from and then that becomes the exact case as you had before. :+1:

scouten commented:
@josevalim @blatyo I am having trouble updating SQLite adapter to this version of Ecto. SQLite has no locks and no `SELECT FOR UPDATE` syntax, so concurrent schema migration is probably not feasible.

Can you advise on a fallback position? Or construct a fallback approach for those of us maintaining adapters for less fancy database engines? :-)

josevalim commented:
@scouten the simplest implementation you can have is this:

```elixir
  def lock_for_migrations(repo, query, _opts, fun) do
    query
    |> repo.all()
    |> fun.()
end
```

Although we should wait for @blatyo's input in case the tools he consulted when implementing this do anything special for sqlite.

blatyo commented:
@scoutin Since SQLite lacks those features, it should define `lock_for_migrations` to fall back back to the previous behavior.

``` elixir
def lock_for_migrations(repo, query, _opts, fun) do
    query
    |> Map.put(:lock, nil)
    |> repo.all()
    |> fun.()
end
```

Basically what @josevalim wrote, but also setting `:lock` to `nil`, since your adapter may use that for something.

scouten commented:
OK, thank you both. Will try that soon.

scouten commented:
Yup. Works. Thank you!

fishcakez commented:
@josevalim the failing alter tests that got removed are very important tests for query type caching we need to bring back equivalent tests.

josevalim commented:
@fishcakez oh, I forgot about them. I have some WIP in a branch, I will push today.

josevalim commented:
@fishcakez 1f26fed3a0b4cd4ba2314fed18d38ae62b7a7ce4

fishcakez commented:
@josevalim 👍

marceldegraaf commented:
@josevalim it looks like this functionality isn't in the `2.2.6` release; do you know when it will be incorporated in a tagged release?

josevalim commented:
It will only be in the next major version.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

marceldegraaf commented:
@josevalim thanks for your reply; you mean it will be available from Ecto `3.x.x` onwards? What would be my best option if I'd like to use this feature already?

josevalim commented:
You need to depend on master.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

marceldegraaf commented:
OK, thanks!
On Tue, 28 Nov 2017 at 20:03, José Valim <notifications@github.com> wrote:

> You need to depend on master.
> --
>
>
> *José Valimwww.plataformatec.com.br
> <http://www.plataformatec.com.br/>Founder and Director of R&D*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-ecto/ecto/pull/2215#issuecomment-347629125>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA6o1JtwnMgumEiuRDOvtjDAYPflyD5ks5s7FkBgaJpZM4PSLk->
> .
>
